Working 1.21 port (for the most part)

SuperKnux opened the pull request.

slava110 committed:
Merge branch 'main' into devel/port-1.21

# Conflicts:
#	Fabric/build.gradle
#	Forge/build.gradle
#	Forge/src/generated/resources/.cache/d4b3e6634c30118e1127c02b727ea285752e5aac
#	Neoforge/src/main/java/at/petrak/hexcasting/forge/loot/ForgeHexCypherLootMod.java

slava110 committed:
Fixed Gradle problems and turned off Fabric temporarily. Will come back later!

slava110 committed:
Ported to 1.21 NeoForge, some Neo-specific stuff is on common package now :c

slava110 committed:
Fixing errors made during port

slava110 committed:
Bumped Java version on workflow from 17 to 21

slava110 committed:
Fix resources

slava110 committed:
Fix inline displaying for entity iotas

slava110 committed:
Fix creative tab item ordering

SuperKnux committed:
* Mostly complete Implementation of Item Components
* Attempted fix of Networking (Incomplete)
* Some minor fixes here and there

SuperKnux committed:
* More attempts to fix networking
* Other fixes here and there (fixed up some whole minor folders so that's pretty cool)
* Fixed NBTUtils!

SuperKnux committed:
* More various fixes like...
* New ContainerInput for Recipes

* Various edits to different Items

* etc, this is setting up for a merge anyway

SuperKnux committed:
Merge remote-tracking branch 'refs/remotes/slava/devel/port-1.21' into indev/1.21.1

SuperKnux committed:
Biiiiiig fixes :3

penguinencounter committed:
that's. that's not how remapping works

SuperKnux committed:
* Mod finally builds, but fails to launch client

SuperKnux committed:
* Replaced serialization-hooks with Fabric API's CustomIngredient system

SuperKnux committed:
* Forgot to remove traces of serialization-hooks

SuperKnux committed:
* New FabricRegister for Registries

SuperKnux committed:
* Attempted fix for Fabric Packet handling

SuperKnux committed:
* Updated loom and fixed Fabric building.

SuperKnux committed:
* Changed some files and mixins :help:

SuperKnux committed:
* Some datagen

SuperKnux committed:
* Finally finished datagen!

SuperKnux committed:
Fixed crash with impeti breaking (also, fixed failure of item stacks visualizing on impeti scrying sight mishap)
Fixed slate rendering both as a block entity and an item stack.

SuperKnux committed:
* Added handlers to S2C Payloads

SuperKnux committed:
Duplicates strategy stuff

SuperKnux committed:
* Adjusted FabricRegister to work correctly, fixes attribute errors on Fabric; might overhaul the registry with it.
* Registered ItemApiLookups (ParticleSprays function on Fabric!!!)
* Readded BrainsweepeeIngredient.getSomeKindOfReasonableIDForEmi(), EMI recipes work for the most part
* Fixed Fabric Creative Tabs (mostly)

SuperKnux committed:
* Turned MobEffects into Holder<MobEffect>s. Might have to do this for other things.
* Fixed Fabric interop naming.

SuperKnux committed:
* Fixed Impeti not spraying particles or making noise when activated (Not when erroring)
* Adjusted Fabric creative mode tab entries (untested)
* Fixed scrolls being unreadable
* Fixed rendering artifacts when both toasts and casting stack are present on screen

SuperKnux committed:
* Fixed slate rendering, scroll rendering, inline pattern rendering (partially)

SuperKnux committed:
* Added a workaround for https://github.com/SamsTheNerd/inline/issues/34
* Fixed Media Cube implementation killing server
* Adjusted neoforge.mods.toml

SuperKnux committed:
* i guess non-flat inline patterns will be fullbright for now :N_Disbelief:

SuperKnux committed:
Update artifact preparation to include Neoforge

SuperKnux committed:
* Fixed particle texture bleeding on Neo
* Fixed ClientCastingStacks not ticking on Neo

SuperKnux committed:
Merge remote-tracking branch 'origin/indev/1.21.1' into indev/1.21.1

SuperKnux committed:
* Removed the hex_unbreakable.jsons in generated data

SuperKnux committed:
* oops

SuperKnux committed:
* Merge updated 1.21 from upstream to indev/1.21.1

SuperKnux committed:
Updated generated resources, added synthetic EMI IDs to Fabric brainsweep recipes, minor fixes on Fabric

object-Object commented on Common/build.gradle:
```diff
@@ -2,8 +2,18 @@ architectury {
+pkSubproj {
+    platform = "common"
+    pkPublish = false
+}
```
This is unnecessary. We configure `platform` in the base `build.gradle`, and `pkPublish` defaults to `false` already.

SuperKnux replied:
Fixed.

object-Object commented on Fabric/build.gradle:
```diff
@@ -142,6 +151,12 @@ dependencies {
+    modLocalRuntime "dev.architectury:architectury-fabric:$architecturyVersion"
+    modLocalRuntime "com.samsthenerd.inline:inline-fabric:$minecraftVersion-$inlineVersion"
+    modLocalRuntime "at.petra-k:paucal:$paucalVersion+$minecraftVersion-fabric"
+    localRuntime "hexcasting:jankson:1.2.3"
```
Surely we can get jankson from an actual maven? Why do we need to vendor it?

SuperKnux replied:
This was the most reliable way via trial and error for me to prevent Fabric:runClient from failing due to a strange NoSuchMethodError related to Jankson; it's in the name :sob: (it randomly shows up or disappears, it's really weird)

object-Object replied:
Yes, but why do we need to include the jar locally? Jankson is on Maven Central. Also, can you put that context in a comment here?

SuperKnux replied:
Changed to Maven Central, added a comment

object-Object commented on Neoforge/build.gradle:
```diff
@@ -72,49 +85,84 @@ repositories {
+    forgeRuntimeLibrary("org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.10.2")
+    forgeRuntimeLibrary("org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:1.10.2")
+    forgeRuntimeLibrary("org.jetbrains.kotlinx:kotlinx-serialization-core-jvm:1.8.1")
+    forgeRuntimeLibrary("org.jetbrains.kotlinx:kotlinx-serialization-json-jvm:1.8.1")
```
this seems weird, I didn't have to do anything like this for my other Architectury mods

SuperKnux replied:
🤷 if I left out any one of these, Neoforge:runClient failed

SuperKnux committed:
IDEA failed to be useful; readded generated resources as they should've existed in the first place

object-Object commented on Common/src/main/resources/assets/hexcasting/textures/block/slate_block.png:
Where'd this come from / why was it added?

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/HexAPI.java:
```diff
@@ -217,10 +184,6 @@ default ArmorMaterial robesMaterial() {
-    /**
-     * Location in the userdata of the number of ops executed
-     */
-    String OP_COUNT_USERDATA = modLoc("op_count").toString();
```
Why'd we get rid of this?

object-Object commented:
A note for future reviewers: you can hide all JSON file changes to make the page work significantly faster

<img width="626" height="289" alt="image" src="https://github.com/user-attachments/assets/e0a6a46e-46ab-4094-a6fb-0fb2de126163" />

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/advancements/MinMaxLongs.java:
```diff
@@ -1,85 +1,81 @@
+        Optional<Long> min,
+        Optional<Long> max,
+        Optional<Long> minSq,
+        Optional<Long> maxSq
```
I'm not a huge fan of this. It's generally good practice to set up APIs such that paradoxical / logically inconsistent states are impossible. For example, here it would be possible to instantiate this record with `min` set but `minSq` empty, which would be very confusing. Is there a specific reason for this to be a record, or could we leave it as a class? Or if it absolutely has to be like this, then we should at least [annotate the primary constructor with ApiStatus.Internal](https://github.com/object-Object/HexDebug/blob/057b8100cdbc35138364c38bee6db9b72c1a2b8e/Common/src/main/java/gay/object/hexdebug/api/splicing/SplicingTableIotaClientView.java#L37-L38).

object-Object replied:
Hmm, gross, but I guess that makes sense; let's just add a comment explaining that, then

SuperKnux replied:
I uh... preemptively removed the min/maxSq stuff. I could probably bring that back but it works as is regardless.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/advancements/OvercastTrigger.java:
```diff
@@ -40,41 +29,25 @@ public void trigger(ServerPlayer player, int mediaGenerated) {
+            MinMaxBounds.Ints mediaGenerated,
+            // This is the *proporttion* of the health bar.
+            MinMaxBounds.Doubles healthUsed,
+            // DID YOU KNOW THERES ONE TO CHECK THE WORLD TIME, BUT NOT THE HEALTH!?
```
thank you for preserving these comments

slava110 replied:
Np, I loved them. Preserved funny comments everywhere where possible I think :p

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/Iota.java:
```diff
@@ -121,7 +134,13 @@ public static boolean tolerates(Iota a, Iota b) {
+        if(o instanceof Iota io) {
+            return tolerates(this, io);
+        }
+        return false;
```
yay!

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/IotaType.java:
```diff
@@ -1,78 +1,52 @@
-        out.put(HexIotaTypes.KEY_DATA, dataTag);
-        return out;
-    }
+    public static final Codec<Iota> TYPED_CODEC = Codec.lazyInitialized(() -> IXplatAbstractions.INSTANCE
```
A note on code style: when you're changing a file, please try to match the style of the code that you're changing. In this case, that means putting abstract methods first, then non-abstract methods, then static fields/methods. It can be confusing to have static, non-static, abstract, etc. methods/fields all intermingled like this - keeping the abstract methods at the top makes it easier for people implementing this class to know what they need to override.

SuperKnux replied:
Did that! :D

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/IotaType.java:
```diff
@@ -1,78 +1,52 @@
-     * Throwing an exception raises a mishap.
-     */
-    @Nullable
-    public abstract T deserialize(Tag tag, ServerLevel world) throws IllegalArgumentException;
```
Why are we removing all of the serialization/deserialization methods?

SuperKnux replied:
Is there an use case that would still require these instead of using Codecs as we have them now? /genq

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/IotaType.java:
```diff
@@ -1,78 +1,52 @@
-     * Get a display of this datum from the {@code data} tag, <i>without</i> the world.
-     * This is for use on the client.
-     */
-    public abstract Component display(Tag tag);
```
Why was this removed? Currently, we absolutely need this in order to display iotas on the client, because in general it's **not possible to deserialize arbitrary iotas on the client**. Has that changed, and if so, how and why?

SuperKnux replied:
I'm going to ignore that last sentence from Slava for the sake of delegating new features/refactoring to AFTER i finish addressing changes. The TYPED_CODEC and TYPED_STREAM_CODEC allow deserialization of arbitrary iotas on the client.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/Iota.java:
```diff
@@ -9,29 +9,38 @@
-    protected final Object payload;
-    @NotNull
-    protected final IotaType<?> type;
+    protected final Supplier<IotaType<? extends Iota>> type;
```
Why does this need to be a `Supplier` now?

object-Object replied:
#1056

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/Iota.java:
```diff
@@ -9,29 +9,38 @@
 
 public abstract class Iota {
     @NotNull
-    protected final Object payload;
```
Why are we removing the payload field?

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/GarbageIota.java:
See NullIota comments, I think they all apply here too

SuperKnux replied:
Fairly sure Garbage is meant to be non-singleton, so I'm reverting this as well.

object-Object commented:
Another note for future reviewers: if you use Bitwarden, apparently disabling its browser extension makes the review page work significantly faster???????

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/ExecutionClientView.kt:
```diff
@@ -11,7 +18,26 @@ data class ExecutionClientView(
 
     // These must be tags so the wrapping of the text can happen on the client
     // otherwise we don't know when to stop rendering
```
Is this comment outdated now?

slava110 replied:
Yep

SuperKnux replied:
Removed.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingImage.kt:
```diff
@@ -1,57 +1,59 @@
+            )
+
             const val TAG_IOTAS = "iotas"
             const val TAG_ESCAPED = "escaped"
```
Do we still need these?

slava110 replied:
No, they are unused

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/env/StaffCastEnv.java:
```diff
@@ -131,7 +132,7 @@ public static void handleNewPatternOnServer(ServerPlayer sender, MsgNewSpellPatt
         IXplatAbstractions.INSTANCE.sendPacketToPlayer(sender,
             new MsgNewSpellPatternS2C(clientInfo, resolvedPatterns.size() - 1));
 
-        IMessage packet;
```
Apple is in shambles 😔

SuperKnux replied:
😔

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/env/CircleCastEnv.java:
```diff
@@ -15,6 +15,7 @@
 import at.petrak.hexcasting.api.pigment.FrozenPigment;
 import at.petrak.hexcasting.common.lib.HexAttributes;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Holder;
```
This seems unnecessary?

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/ContinuationFrame.kt:
```diff
@@ -50,41 +57,36 @@ interface ContinuationFrame {
+        fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, U>
+        
+        companion object {
+            //TODO if doesn't exist use `FrameEvaluate(SpellList.LList(0, listOf()), false)` as default value
```
What does this mean?

SuperKnux replied:
@slava110 ?

object-Object replied:
#1060

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/ContinuationFrame.kt:
```diff
@@ -50,41 +57,36 @@ interface ContinuationFrame {
-            return (tag.get(HexContinuationTypes.KEY_DATA) as? CompoundTag)?.let { type.deserializeFromNBT(it, world) }
-                ?: FrameEvaluate(SpellList.LList(0, listOf()), false)
-        }
+        fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, U>
```
Could we change these to abstract fields instead of abstract functions? It'll make the Kotlin subclass implementations a lot cleaner.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/SpellContinuation.kt:
```diff
@@ -16,31 +19,43 @@ sealed interface SpellContinuation {
-        const val TAG_FRAME = "frame"
-
+        // TODO port: maybe serialize to list like before?
+        // TODO port: maybe unit should be first
```
What do these mean?

slava110 replied:
This was something very cursed and I wasn't sure if it works or not. Still not sure

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/mishaps/MishapAlreadyBrainswept.kt:
```diff
@@ -14,11 +13,7 @@ class MishapAlreadyBrainswept(val mob: Mob) : Mishap() {
-            (mob.damageSources() as AccessorDamageSource).`hex$source`(HexDamageTypes.OVERCAST, ctx.castingEntity)!!,
-            mob.health
-        )
+        mob.hurt(mob.damageSources().source(HexDamageTypes.OVERCAST, ctx.castingEntity), mob.health)
```
What's the difference between `trulyHurt` and `mob.hurt`?

SuperKnux replied:
As explained by Robotgiggle, `trulyHurt` attempts to bypass mods that use invasive mixins to prevent damage.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/mishaps/MishapBadEntity.kt:
```diff
@@ -19,7 +19,7 @@ class MishapBadEntity(val entity: Entity, val wanted: Component) : Mishap() {
 
     override fun errorMessage(ctx: CastingEnvironment, errorCtx: Context) =
-        error("bad_entity", wanted, entity.displayName!!.plainCopy().aqua)
+        error("bad_entity", wanted, entity.displayName?.plainCopy()?.aqua ?: entity.type.description.plainCopy().aqua)
```
This could be refactored, eg. `(entity.displayName ?: entity.type.description).plainCopy().aqua`

SuperKnux replied:
Refactored!

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/mishaps/MishapEntityTooFarAway.kt:
```diff
@@ -16,5 +17,5 @@ class MishapEntityTooFarAway(val entity: Entity) : Mishap() {
 
     override fun errorMessage(ctx: CastingEnvironment, errorCtx: Context): Component =
-        error("entity_too_far", entity.displayName!!)
+        error("entity_too_far", entity.displayName?.plainCopy() ?: entity.type.description)
```
Is there a plainCopy missing for the description?

SuperKnux replied:
Refactored.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/mod/HexTags.java:
```diff
@@ -54,6 +54,8 @@ public static final class Blocks {
 
         public static final TagKey<Block> CHEAP_TO_BREAK_BLOCK = create("cheap_to_break_block");
 
+        public static final TagKey<Block> HEX_UNBREAKABLE = create("hex_unbreakable");
```
Can you add a comment explaining what this is for?

slava110 replied:
I think something was removed between versions and I replaced it, but I don't remember exactly

SuperKnux replied:
Done

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/utils/NBTHelper.kt:
This file has a bunch of unnecessary formatting changes, can we try to minimize those?

SuperKnux replied:
elaborate for me please?

object-Object commented on Common/src/main/java/at/petrak/hexcasting/client/entity/WallScrollRenderer.java:
What are these changes for?

SuperKnux replied:
@slava110

object-Object commented on Common/src/main/java/at/petrak/hexcasting/client/gui/GuiSpellcasting.kt:
```diff
@@ -40,7 +43,7 @@ import kotlin.math.*
     private val handOpenedWith: InteractionHand,
     private var patterns: MutableList<ResolvedPattern>,
-    private var cachedStack: List<CompoundTag>,
+    private var cachedStack: List<Iota>,
     private var cachedRavenmind: CompoundTag?,
```
Same question as the other place I asked it - why the difference in type here?

SuperKnux replied:
@slava110

object-Object replied:
#1059

object-Object commented on Common/src/main/java/at/petrak/hexcasting/client/render/WorldlyPatternRenderHelpers.java:
```diff
@@ -52,7 +54,7 @@ public static void renderPatternForScroll(HexPattern pattern, EntityWallScroll s
         ps.translate(-blockSize / 2f, -blockSize / 2f, 1f / 32f);
         // there's almost certainly a better way to do this, but we're just flipping the y and z axes to fix normals
-        ps.last().normal().mul(new Matrix3f(1, 0, 0, 0, 0, 1, 0, 1, 0));
+        // no need for this anymore :D
```
can we remove the comment, then?

SuperKnux replied:
done

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/blocks/akashic/BlockEntityAkashicBookshelf.java:
```diff
@@ -70,23 +81,25 @@ public void clearIota() {
-            compoundTag.put(TAG_IOTA, this.iotaTag);
+    protected void saveModData(CompoundTag compoundTag, HolderLookup.Provider registries) {
+        if (this.pattern != null && this.iota != null) {
+            compoundTag.put(TAG_PATTERN, HexPattern.CODEC.encodeStart(NbtOps.INSTANCE, pattern).getOrThrow());
```
This is a pretty arbitrary place to put this comment, but I think it would make addons a lot easier to port if we could keep the serialize/deserialize methods and just change the implementations to use codecs. Open for reasons not to do this though.

slava110 replied:
I'm kinda opposed to keeping old code just to make porting easier. Addons will still need to change a lot of code to be compatible with 1.21 at all. By leaving old methods for compatibility you lead yourself into Bukkit pitfall where some methods from 1.6.4 use a lot of hacks nowadays just to keep compatibility with old plugins.
For 1.20.1 changes I'd completely agree. But it's a between-mc-versions port afterall, I think that keeping redundant code isn't good for that

SuperKnux replied:
Pretty much what Slava said in regards to this. Codecs are pretty much REQUIRED for things using Item Components to work.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/blocks/circles/impetuses/BlockRightClickImpetus.java:
```diff
@@ -17,21 +20,30 @@ public BlockRightClickImpetus(Properties p_49795_) {
             if (tile instanceof BlockEntityRightClickImpetus impetus) {
-                if (pPlayer instanceof ServerPlayer serverPlayer) {
+                if (player instanceof ServerPlayer sPlayer) {
 //                    impetus.activateSpellCircle(serverPlayer);
```
While we're here, can we remove this?

SuperKnux replied:
removed the comment

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/casting/actions/local/OpPeekLocal.kt:
```diff
@@ -9,15 +9,16 @@ import at.petrak.hexcasting.api.casting.eval.vm.SpellContinuation
 
         val rm = if (image.userData.contains(HexAPI.RAVENMIND_USERDATA)) {
-            IotaType.deserialize(image.userData.getCompound(HexAPI.RAVENMIND_USERDATA), env.world)
+            IotaType.TYPED_CODEC.parse(NbtOps.INSTANCE, image.userData.getCompound(HexAPI.RAVENMIND_USERDATA)).getOrThrow()
```
I don't think `getOrThrow` is what we want here. The previous behaviour on failure was to return garbage.

SuperKnux replied:
`getOrThrow` with the IotaType `TYPED_CODEC` results in a successful DataResult because of its' `orElse` method, this should work.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/casting/actions/spells/OpColorize.kt:
Unnecessary whitespace changes

SuperKnux replied:
Removed

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/casting/actions/spells/OpDestroyFluid.kt:
```diff
@@ -79,7 +79,7 @@ object OpDestroyFluid : SpellAction {
                         ) {
                             val success =
                                 if (blockstate.block is BucketPickup && !(blockstate.block as BucketPickup).pickupBlock(
-                                        env.caster,
```
Nice catch

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/casting/actions/spells/OpIgnite.kt:
```diff
@@ -84,7 +77,9 @@ object OpIgnite : SpellAction {
     private data class EntitySpell(val entity: Entity) : RenderedSpell {
         override fun cast(env: CastingEnvironment) {
-            entity.igniteForSeconds(8f)
+            if (!entity.fireImmune()) {
```
Should we check this ahead of time and mishap?

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/entities/EntityWallScroll.java:
```diff
@@ -86,13 +85,17 @@ public void setShowsStrokeOrder(boolean b) {
-    @Override
-    public int getHeight() {
-        return 16 * blockSize;
+    protected AABB calculateBoundingBox(BlockPos pos, Direction p_direction) {
```
What's this change for?

SuperKnux replied:
<img width="1120" height="351" alt="image" src="https://github.com/user-attachments/assets/04519666-f47f-4a51-ab8e-f9bcae361f37" />
@slava110 idk about that emoji
Though, I had did the same thing pre-merge due to Minecraft making a method privated I think.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/items/ItemStaff.java:
```diff
@@ -43,10 +45,14 @@ public InteractionResultHolder<ItemStack> use(Level world, Player player, Intera
+            var userData = vm.getImage().getUserData();
+            CompoundTag ravenmind = null;
+            if(userData.contains(HexAPI.RAVENMIND_USERDATA))
+                ravenmind = userData.getCompound(HexAPI.RAVENMIND_USERDATA);
```
I see that we're doing basically this same thing in several places. Can we add something like a `getRavenmind` helper to `CastingImage`?

SuperKnux replied:
Done! `Optional<Tag>` `ravenmind()` method added!

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/items/magic/ItemCreativeUnlocker.java:
```diff
@@ -244,17 +236,18 @@ public ItemStack finishUsingItem(ItemStack stack, Level level, LivingEntity cons
         }
 
         return component.withStyle((s) -> s.withColor(
             TextColor.fromRgb(Mth.hsvToRgb((level.getGameTime() + shift) * 2 % 360 / 360F, 1F, 1F))));
```
Can we refactor this to have less duplicated code?

SuperKnux replied:
Elvis operator my beloved; fixed

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/misc/ContainerInput.kt:
```diff
@@ -0,0 +1,21 @@
+import net.minecraft.world.item.ItemStack
+import net.minecraft.world.item.crafting.RecipeInput
+
+abstract class ContainerInput<C : Container>(val container: Container) : RecipeInput {
```
Can you add a doc comment describing what this is for?

SuperKnux replied:
Actually, it's being removed for the same reasons as `HexBlockFallingBlock` and `HexItemComponents`.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/recipe/ingredient/brainsweep/BrainsweepeeIngredients.java:
```diff
@@ -0,0 +1,105 @@
+import java.util.List;
+import java.util.function.BiConsumer;
+
+public class BrainsweepeeIngredients {
```
This seems like a registration file - shouldn't this be in `lib`?

slava110 replied:
Could be in lib, yep

SuperKnux replied:
Moved to `lib`, renamed to `HexBrainsweepeeIngredients`.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/recipe/ingredient/state/StateIngredients.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Random;
+import java.util.function.BiConsumer;
+
+public class StateIngredients {
```
This seems like a registration file, should it be in `lib`?

SuperKnux replied:
moved to `lib`, renamed to `HexStateIngredients`.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/recipe/ingredient/state/StateIngredients.java:
```diff
@@ -0,0 +1,95 @@
+    };
+
+    public static void register(BiConsumer<StateIngredientType<?>, ResourceLocation> r) {
+        r.accept(NONE_TYPE, HexAPI.modLoc("none"));
+        r.accept(BLOCK_TYPE, HexAPI.modLoc("block"));
+        r.accept(BLOCK_STATE, HexAPI.modLoc("state"));
+        r.accept(BLOCKS, HexAPI.modLoc("blocks"));
+        r.accept(TAG, HexAPI.modLoc("tag"));
+        r.accept(TAG_EXCLUDING, HexAPI.modLoc("tag_excluding"));
```
This doesn't match our convention for how to implement this

slava110 replied:
You are free to modify it to match your convention I guess 🤷

SuperKnux replied:
🤷‍♀️

object-Object replied:
#1065

object-Object commented on Common/src/main/java/at/petrak/hexcasting/datagen/recipe/HexplatRecipes.java:
```diff
@@ -594,12 +599,154 @@ private void gayRecipe(Consumer<FinishedRecipe> recipes, ItemPridePigment.Type t
+        SpecialRecipeBuilder.special(recipeFunc::apply).save(consumer, HexAPI.MOD_ID + ":dynamic" + name.getPath());
+    }
+
+    protected static Criterion<InventoryChangeTrigger.TriggerInstance> hasItem(ItemLike p_125978_) {
```
Name

SuperKnux replied:
renamed to `itemLike`.

object-Object commented on Common/src/main/java/at/petrak/hexcasting/datagen/recipe/HexplatRecipes.java:
```diff
@@ -594,12 +599,154 @@ private void gayRecipe(Consumer<FinishedRecipe> recipes, ItemPridePigment.Type t
+        );
+    }
+
+    // ================================= From PAUCAL 1.20
```
Why were these moved out of PAUCAL? These are useful helpers for addons - can we put them back in PAUCAL?

object-Object replied:
#1070

object-Object commented on Common/src/main/java/at/petrak/hexcasting/xplat/IClientXplatAbstractions.java:
```diff
@@ -40,6 +40,8 @@ public interface IClientXplatAbstractions {
 
     IClientXplatAbstractions INSTANCE = find();
 
+    String getModelLocVariant();
```
Please add a doc comment

SuperKnux replied:
Done!

object-Object commented on Common/src/main/java/at/petrak/hexcasting/xplat/IXplatRegister.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.function.Supplier;
+
+/**
+ * Handles registration
```
Can you elaborate a bit?

SuperKnux replied:
Added a better doc for IXplatRegister.

object-Object commented on Fabric/src/main/java/at/petrak/hexcasting/fabric/FabricHexInitializer.kt:
```diff
@@ -59,14 +99,22 @@ object FabricHexInitializer : ModInitializer {
+                }
+            }
+        }
+
```
What's this for?

object-Object reviewed (changes requested):
Mostly only reviewed `Common`, will come back and look through the platform code later. Left many questions/comments/suggestions.

I think my main concern is with how EntityIota was ported. I'm not sure how we want it to work, but I'm not satisfied with how it currently is. We can discuss it more on Discord.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/xplat/IXplatRegister.java:
```diff
@@ -0,0 +1,21 @@
+ * Handles registration
+ * @param <B>
+ */
+public interface IXplatRegister<B> {
```
it was like that when i got here :sob:, i can change it to T instead

slava110 replied:
It's `B` because it is `BASE` class for registry entry.
`MyItem extends Item`, where `MyItem` is `T` generic parameter in `register` method of this class and `B` is base class of registry, type parameter of the class

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/advancements/OvercastTrigger.java:
```diff
@@ -40,41 +29,25 @@ public void trigger(ServerPlayer player, int mediaGenerated) {
+                                MinMaxBounds.Ints.CODEC.fieldOf("media_generated").forGetter(Instance::mediaGenerated),
+                                MinMaxBounds.Doubles.CODEC.fieldOf("health_used").forGetter(Instance::healthUsed),
+                                // HEY KIDS DID YOYU KNOW THERE'S NOT A CRITERIA FOR HOW MUCH ***HEALTH*** AN ENTITY HAS
+                                MinMaxBounds.Doubles.CODEC.fieldOf("mojang_i_am_begging_and_crying_please_add_an_entity_health_criterion").forGetter(Instance::healthLeft)
```
### yep XD

SuperKnux commented on Common/build.gradle:
```diff
@@ -25,17 +35,25 @@ repositories {
 
+loom {
+    accessWidenerPath = file("src/main/resources/hexplat.accesswidener")
+}
```
Fixed.

slava110 commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/NullIota.java:
```diff
@@ -13,15 +16,14 @@
  * An iota with no data associated with it.
  */
 public class NullIota extends Iota {
+    public static final NullIota INSTANCE = new NullIota();
```
- I didn't see any reason to have multiple instance of same object. Most straightforward optimization IMO
- These addons probably require mixins because in 1.20.1 base hex casting null iotas have the same hash code due to singleton payload
- Could be done for boolean as well, yeah

SuperKnux replied:
NullIota has had its' old behavior reinstated for the time being.

slava110 commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/ListIota.java:
```diff
@@ -102,46 +93,49 @@ public int depth() {
+            // only add a comma between 2 non-patterns (commas don't look good with Inline patterns)
+            // TODO: maybe add a method on IotaType to allow it to opt out of commas?
+            if (i < list.size() - 1) {
+                var thisIotaNeedsComma = sub.type != PatternIota.TYPE;
```
Probably doesn't, good catch

SuperKnux replied:
caught it, edited

slava110 commented on Common/src/main/java/at/petrak/hexcasting/api/casting/ActionUtils.kt:
```diff
@@ -32,10 +34,28 @@ fun List<Iota>.getDouble(idx: Int, argc: Int = 0): Double {
+        return x.cachedEntity;
     } else {
         throw MishapInvalidIota.ofType(x, if (argc == 0) idx else argc - (idx + 1), "entity")
     }
```
Unloadable entities, discussed in Discord

SuperKnux replied:
They are no longer there as EntityIota's old behavior has been restored.

slava110 commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingImage.kt:
```diff
@@ -72,60 +74,32 @@ data class CastingImage private constructor(
-        const val TAG_OPS_CONSUMED = "ops_consumed"
-        const val TAG_USERDATA = "userdata"
-
         @JvmStatic
```
Didn't know about this annotation, useful to know, thanks

object-Object replied:
#1057

slava110 commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/NullIota.java:
```diff
@@ -13,15 +16,14 @@
  */
 public class NullIota extends Iota {
+    public static final NullIota INSTANCE = new NullIota();
     private static final Object NULL_SUBSTITUTE = new Object();
```
Yeah, probably

SuperKnux replied:
Removed!

slava110 commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/ContinuationFrame.kt:
```diff
@@ -98,9 +100,11 @@ interface ContinuationFrame {
+                ?: return null
 
-            return typeLoc?.let { HexContinuationTypes.REGISTRY[it] }
+            return HexContinuationTypes.REGISTRY[typeLoc]
```
Thanks. I think it's a bit worse performance-wise but eh. Doesn't matter probably

slava110 commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/SpellContinuation.kt:
```diff
@@ -16,31 +19,43 @@ sealed interface SpellContinuation {
+            ).map<T>(
+                Function { either: Either<T, T> -> Either.unwrap(either) },
+                Function { value: T -> Either.left(value) }
+            )
```
If it needs to be reused - sure. I just didn't see any other place in the code where it would be applicable

slava110 commented on Common/src/main/java/at/petrak/hexcasting/datagen/recipe/HexplatRecipes.java:
```diff
@@ -455,72 +460,72 @@ public void buildRecipes(Consumer<FinishedRecipe> recipes) {
 
         // Create compat
-        this.conditions.apply(new CreateCrushingRecipeBuilder()
+        /*this.conditions.apply(new CreateCrushingRecipeBuilder()
```
Create 1.21 didn't exist at the time of my port I think

SuperKnux replied:
It appears that it straight up *won't* exist on fabric for 1.21 as Fabricators are skipping it, so this will have to be Neo specific code.

SuperKnux replied:
actually, I think it'll be fine just staying up, since it would only be applied on Neo and silently fail on Fabric.

object-Object replied:
#1063

slava110 commented on Common/src/main/java/at/petrak/hexcasting/common/recipe/ingredient/state/StateIngredientBlock.java:
Because it has more types now? I don't remember tbh

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/circles/BlockEntityAbstractImpetus.java:
```diff
@@ -373,7 +375,7 @@ protected int semitoneFromScale(int note) {
     // this is a good use of my time
     private static final int[] MAJOR_SCALE = {0, 2, 4, 5, 7, 9, 11, 12};
-    private static final int[] MINOR_SCALE = {0, 2, 3, 5, 7, 8, 11, 12};
+    private static final int[] MINOR_SCALE = {0, 2, 3, 5, 7, 8, 10, 12};
```
Matches a proper B Minor scale now

SuperKnux commented on Fabric/src/main/java/at/petrak/hexcasting/fabric/FabricHexInitializer.kt:
```diff
@@ -108,30 +156,67 @@ object FabricHexInitializer : ModInitializer {
+                    HexTags.Actions.PER_WORLD_PATTERN
+                )
+            ) keyList.add(key)
+            keyList.sortWith(Comparator.comparing<ResourceKey<ActionRegistryEntry?>?, ResourceLocation?>(Function { obj: ResourceKey<ActionRegistryEntry?>? -> obj!!.location() }))
```
I think I can just remove the safe calls and it'll be fine

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/xplat/IXplatAbstractions.java:
```diff
@@ -180,12 +181,18 @@ <T extends BlockEntity> BlockEntityType<T> createBlockEntityType(BiFunction<Bloc
 
     boolean isPlacingAllowed(ServerLevel world, BlockPos pos, ItemStack blockStack, @Nullable Player player);
 
+    <B> IXplatRegister<B> createRegistar(ResourceKey<Registry<B>> registryKey);
```
Added.

SuperKnux commented on Fabric/build.gradle:
```diff
@@ -9,6 +10,7 @@ architectury {
 
 // FIXME: update dependencies/versions before releasing 0.12
 pkSubproj {
+    platform = "fabric"
```
Fixed.

SuperKnux commented on build.gradle:
```diff
@@ -144,8 +145,19 @@ allprojects {
+        jvmTarget = JvmTarget.JVM_21
+    }
+}
+
```
Fixed.

SuperKnux commented on Neoforge/gradle.properties:
```diff
@@ -1,10 +1,13 @@
-kotlinForForgeVersion=5.6.0
+kotlinForForgeVersion=5.8.0
+kotlinBaseVersion=2.1.20
+kotlinxCoroutinesVersion=
```
Forgot to remove that

SuperKnux commented on Common/src/generated/resources/data/neoforge/loot_modifiers/global_loot_modifiers.json:
Don't really know what this is doing here or why... 😅

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/IotaType.java:
```diff
@@ -88,99 +62,10 @@ private static boolean isTooLargeToSerialize(Iterable<Iota> examinee, int starti
-        }
-        return type.color();
-    }
+    public abstract int color();
```
Done.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/BooleanIota.java:
```diff
@@ -32,21 +37,30 @@ public boolean toleratesOther(Iota that) {
+
+    @Override
+    public Component display() {
+        return BooleanIota.display(getBool());
```
Done.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/NullIota.java:
```diff
@@ -35,20 +37,28 @@ public boolean toleratesOther(Iota that) {
-    public @NotNull Tag serialize() {
-        return new CompoundTag();
+    public int hashCode() {
+        return NULL_SUBSTITUTE.hashCode();
```
I'll try setting its' hashCode to 0 for now. It's a pretty small change so it's easy to revert if it doesn't work.

SuperKnux committed:
Start of changes to comply with PR requested changes;
Reverted NullIota and GarbageIota implementations,
Adjusted various things,
Got FabricRegister working,
etc

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/EntityIota.java:
I was able to reinstate EntityIotas' old behavior with a little trick Miyu showed me.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/ParticleSpray.kt:
```diff
@@ -23,6 +39,8 @@ data class ParticleSpray(val pos: Vec3, val vel: Vec3, val fuzziness: Double, va
         }
     }
 
+
+
```
mb

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/addldata/ADIotaHolder.java:
Will do!

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/SpellList.kt:
```diff
@@ -90,4 +92,17 @@ sealed class SpellList : Iterable<Iota> {
+    companion object {
+        @JvmStatic
+        val CODEC = IotaType.TYPED_CODEC.listOf().xmap(
+            { list -> LList(list) as SpellList },
```
It seems that FrameEvaluate complains about it otherwise.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/mishaps/MishapEntityUnavailable.kt:
```diff
@@ -0,0 +1,20 @@
+import net.minecraft.world.item.DyeColor
+import java.util.*
+
+class MishapEntityUnavailable(val entityId: UUID, val entityName: Component?) : Mishap() {
```
I have documentation, but it will not be added because it seems like this will not be a thing.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/item/PigmentItem.java:
if you mean the whitespace, I removed it

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/iota/DoubleIota.java:
```diff
@@ -2,23 +2,30 @@
 
     public double getDouble() {
-        return HexUtils.fixNAN((Double) this.payload);
+        return HexUtils.fixNAN((Double) value);
```
IDEA is saying it's unnecessary, so...

SuperKnux committed:
Addressing more changes;
Added new MishapEntityNotFound mishap documentation in Patchouli, among other things

SuperKnux committed:
Also, optimizing imports

SuperKnux committed:
Old EntityIota behavior restored

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/ExecutionClientView.kt:
```diff
@@ -11,7 +18,26 @@ data class ExecutionClientView(
-    val stackDescs: List<CompoundTag>,
-    val ravenmind: CompoundTag?,
-)
+    val stackDescs: List<Iota>,
+    val ravenmind: CompoundTag?
```
Seems to work? Do I have to change this back?

object-Object replied:
#1059

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/client/render/HexAdditionalRenderers.java:
```diff
@@ -123,7 +124,7 @@ private static void renderSentinel(Sentinel sentinel, LocalPlayer owner,
             v.accept(bottom, Icos.TOP_RING[(i + 3) % 5]);
         }
-        tess.end();
+        //tess.end();
```
@slava110

object-Object replied:
#1063

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/blocks/decoration/BlockHexFallingBlock.java:
This was back when my fork was on its' own before being merged; it has been removed.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/lib/HexItemComponents.java:
Same reason as `HexBlockFallingBlock`. Removed.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/interop/inline/InlinePatternRenderer.java:
```diff
@@ -61,7 +62,8 @@ public int render(InlinePatternData data, GuiGraphics drawContext, int index, St
         int color = trContext.usableColor();
-        PatternRenderer.renderPattern(data.pattern, drawContext.pose(), new PatternRenderer.WorldlyBits(drawContext.bufferSource(), trContext.light(), null),
+        boolean isFlat = InlineRenderer.isFlat(drawContext.pose(), trContext.layerType());
+        PatternRenderer.renderPattern(data.pattern, drawContext.pose(), !isFlat ? new PatternRenderer.WorldlyBits(drawContext.bufferSource(), LightTexture.FULL_BRIGHT, null) : null,
```
Attempts to fix patterns rendering strangely on signs, i.e. too dark or too bright. These attempts have failed so far so I'd like suggestions on your side if that's okay.

object-Object replied:
#1071

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/casting/actions/local/OpPushLocal.kt:
```diff
@@ -22,7 +24,7 @@ object OpPushLocal : Action {
             image.userData.remove(HexAPI.RAVENMIND_USERDATA)
          else
-            image.userData.put(HexAPI.RAVENMIND_USERDATA, IotaType.serialize(newLocal))
+            image.userData.put(HexAPI.RAVENMIND_USERDATA, IotaType.TYPED_CODEC.encodeStart(NbtOps.INSTANCE, newLocal).getOrThrow())
```
See `OpPeekLocal`.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/command/PatternResLocArgument.java:
```diff
@@ -20,6 +20,7 @@
 import java.util.ArrayList;
 import java.util.concurrent.CompletableFuture;
 
+@Deprecated
```
Done

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/ResolvedPattern.kt:
```diff
@@ -2,26 +2,40 @@ package at.petrak.hexcasting.api.casting.eval
+            Codec.INT.fieldOf("OriginR").forGetter { it.origin.r },
+            Codec.STRING.xmap(
+                { ResolvedPatternType.fromString(it) },
+                { it.name.lowercase() }
```
Removed lowercase.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/api/casting/eval/ResolvedPattern.kt:
```diff
@@ -2,26 +2,40 @@ package at.petrak.hexcasting.api.casting.eval
-        }
+        @JvmField
+        val CODEC: Codec<ResolvedPattern> = RecordCodecBuilder.create({instance -> instance.group(
+            HexPattern.CODEC.fieldOf("pattern").forGetter { it.pattern },
```
Changed

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/items/storage/ItemAbacus.java:
```diff
@@ -52,13 +52,13 @@ public void writeDatum(ItemStack stack, Iota datum) {
 
             var key = "hexcasting.tooltip.abacus.reset";
-            if (oldNum == 69) {
+            if (oldNum != null && oldNum == 69) {
```
nice

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/lib/HexDataComponents.java:
Made a couple changes. Feel free to suggest changes.

object-Object replied:
#1064

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/xplat/IXplatAbstractions.java:
```diff
@@ -127,7 +128,7 @@ public interface IXplatAbstractions {
     ADHexHolder findHexHolder(ItemStack stack);
 
-    @Nullable ADVariantItem findVariantHolder(ItemStack stack);
+    ADVariantItem findVariantHolder(ItemStack stack);
```
I dunno why I didn't make this Nullable, good catch. Fixed.

SuperKnux committed:
Made a few more changes as well as refactoring, ready for review

SuperKnux committed:
Fixed patterns rendering too dark on signs, attempted to fix patterns rendering too dark on slates... again

SuperKnux commented:
Waiting for review.

SuperKnux committed:
chore: remove obsolete ExecutionClientView comment

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/loot/AddHexToAncientCypherFunc.java:
```diff
@@ -36,7 +35,12 @@
 public class AddHexToAncientCypherFunc extends LootItemConditionalFunction {
-    public AddHexToAncientCypherFunc(LootItemCondition[] lootItemConditions) {
+    public static final MapCodec<AddHexToAncientCypherFunc> CODEC = RecordCodecBuilder.mapCodec(
+            p_344674_ -> commonFields(p_344674_)
```
Done.

SuperKnux commented on Common/src/main/java/at/petrak/hexcasting/common/loot/AmethystReducerFunc.java:
```diff
@@ -1,25 +1,38 @@
 public class AmethystReducerFunc extends LootItemConditionalFunction {
-    public final double delta;
+    public static final MapCodec<AmethystReducerFunc> CODEC = RecordCodecBuilder.mapCodec(
+            p_344674_ -> commonFields(p_344674_)
```
Done.

SuperKnux committed:
chore: remove obsolete CastingImmage variables

SuperKnux committed:
fix(rendering): restore worldly pattern rendering behavior from 1.20

SuperKnux committed:
chore: changed push branch to 'indev/1.21.1' in PR workflow

Updated the push branch in the PR workflow from 'main' to 'indev/1.21.1'.

SuperKnux committed:
fix(workflow): change dependencies for pr workflow

Updated dependencies and modloader configurations for neoforge.

SuperKnux committed:
fix(rendering): fix pattern inner lines disappearing on uis

SuperKnux committed:
fix(casting): fix ravenmind optional never being null

object-Object commented on Fabric/src/main/java/at/petrak/hexcasting/fabric/cc/adimpl/CCHexHolder.java:
```diff
@@ -59,5 +61,15 @@ public void clearHex() {
             return this.hexHolder.getPigment(this.stack);
         }
+
+        @Override
+        public void readFromNbt(CompoundTag tag, HolderLookup.Provider registryLookup) {
+
+        }
+
+        @Override
+        public void writeToNbt(CompoundTag tag, HolderLookup.Provider registryLookup) {
+
+        }
```
Are these supposed to be empty? Do we need to add a TODO comment or something?

object-Object replied:
#1073

object-Object commented on Fabric/src/main/java/at/petrak/hexcasting/fabric/cc/HexCardinalComponents.java:
```diff
@@ -85,35 +89,18 @@ public void registerEntityComponentFactories(EntityComponentFactoryRegistry regi
 
-        registry.register(i -> i instanceof IotaHolderItem, IOTA_HOLDER, CCItemIotaHolder.ItemBased::new);
+        registry.registerMigration(modLoc("iota_holder"), HexDataComponents.IOTA_HOLDER_IOTA);
         // oh havoc, you think you're so funny
         // the worst part is you're /right/
```
Should this comment be removed (or moved to FabricHexInitializer) since the code it was commenting on is no longer here?

object-Object replied:
#1076

object-Object commented on Neoforge/src/main/java/at/petrak/hexcasting/forge/network/ForgePacketHandler.java:
```diff
@@ -1,80 +1,67 @@
-        int messageIdx = 0;
+    public static void init(IEventBus modBus) {
+        modBus.addListener(RegisterPayloadHandlersEvent.class, ev -> {
+            final PayloadRegistrar registar = ev.registrar("1");
```
Typo: "registar" should be "registrar"

object-Object replied:
#1075

object-Object reviewed (approved):
None of the remaining comments are blocking; I think I'm ok with merging this pretty much as-is and addressing remaining issues in followup PRs. Thanks everyone for your work on this!

SuperKnux committed:
chore: remove unneeded jankson dep on Fabric; temporary flatdir PAUCAL jars
fix(rendering): replace pattern rendering depth with VCDrawHelper depth reversal

SuperKnux committed:
chore: remove innerZ comment

SuperKnux committed:
fix(workflow): attempted fix to test part of pr workflow

SuperKnux committed:
fix(workflow): update headlesshq versions in pr workflow

SuperKnux committed:
fix(workflow): update Fabric API, add more Neo dependencies

SuperKnux committed:
fix(workflow): ACTUALLY update Fabric API in pr workflow

SuperKnux committed:
fix(workflow): change server java version

this is getting annoying...

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/circles/CircleExecutionState.java:
```diff
@@ -171,48 +168,54 @@ public CompoundTag save() {
         out.put(TAG_CURRENT_POS, NbtUtils.writeBlockPos(this.currentPos));
         out.putByte(TAG_ENTERED_FROM, (byte) this.enteredFrom.ordinal());
-        out.put(TAG_IMAGE, this.currentImage.serializeToNbt());
+        out.put(TAG_IMAGE, CastingImage.getCODEC().encodeStart(NbtOps.INSTANCE, currentImage).getOrThrow());
```
#1061

object-Object commented on Common/src/main/java/at/petrak/hexcasting/api/casting/mishaps/MishapOthersName.kt:
```diff
@@ -33,14 +34,19 @@ class MishapOthersName(val confidant: Player) : Mishap() {
-                if (datumToCheck is EntityIota && datumToCheck.entity is Player && datumToCheck.entity != caster)
-                    return datumToCheck.entity as Player
+
+                if(datumToCheck is EntityIota) {
+                    var ent = datumToCheck.getOrFindEntity(level)
+                    if(ent is Player && ent != caster)
+                        return ent
+                }
```
#1062

object-Object commented on Common/src/main/java/at/petrak/hexcasting/common/blocks/akashic/BlockEntityAkashicBookshelf.java:
```diff
@@ -35,14 +39,20 @@ public HexPattern getPattern() {
+        return iota;
     }
 
+    /*@Nullable
+    public Tag getIotaTag() {
+        return iotaTag;
+    }*/
```
#1063

object-Object commented on Common/src/main/java/at/petrak/hexcasting/datagen/recipe/HexplatRecipes.java:
```diff
@@ -455,72 +460,72 @@ public void buildRecipes(Consumer<FinishedRecipe> recipes) {
-            .unlockedBy("enlightenment", enlightenment)
-            .save(recipes, modLoc("brainsweep/budding_amethyst"));
+            .unlockedBy("enlightenment", new Criterion<>(HexAdvancementTriggers.OVERCAST_TRIGGER, enlightenment))
+            .save(recipes, modLoc("budding_amethyst"));
```
#1069
